pallets/parallel-xcm/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+pub type BalanceOf<T> =
+    <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Assets for deposit/withdraw assets to/from crowdloan account
+        type Assets: Transfer<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Inspect<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Mutate<Self::AccountId, AssetId = CurrencyId, Balance = Balance>;
+
+        /// XCM message sender
+        type XcmSender: SendXcm;
+
+        /// Relay network
+        #[pallet::constant]
+        type RelayNetwork: Get<NetworkId>;
+
+        /// The block number provider
+        type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn xcm_fees)]
+    pub type XcmFees<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn total_reserves)]
+    pub type TotalReserves<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Xcm message send failure
+        SendXcmError,
+    }
+}
+
+pub trait ParallelXCM<Balance, AssetId, AccountId> {
+    fn update_xcm_fees(fees: Balance);
+
+    fn update_total_reserves(reserves: Balance) -> DispatchResult;
+
+    fn ump_transact(
+        call: DoubleEncoded<()>,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+    ) -> Result<Xcm<()>, DispatchError>;
+
+    fn do_withdraw(
+        para_id: ParaId,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        para_account_id: AccountId,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+    ) -> Result<(), DispatchError>;
+
+    fn do_contribute(
+        para_id: ParaId,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        amount: Balance,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+        who: &AccountId,
+    ) -> Result<(), DispatchError>;
+}
+
+impl<T: Config> ParallelXCM<BalanceOf<T>, AssetIdOf<T>, T::AccountId> for Pallet<T> {
+    fn update_xcm_fees(fees: BalanceOf<T>) {
+        XcmFees::<T>::mutate(|v| *v = fees);
+    }
+
+    fn update_total_reserves(reserves: BalanceOf<T>) -> DispatchResult {
+        TotalReserves::<T>::try_mutate(|b| -> DispatchResult {
+            *b = b.checked_add(reserves).ok_or(ArithmeticError::Overflow)?;
+            Ok(())
+        })
+    }
```

Comment:
2/3 `TotalReserves` updated here, however, the asset actually transferred outside.

Reply:
same as above